File: tests/rpc/initialize.ts
import * as anchor from "@project-serum/anchor";
import {
    SystemProgram,
} from '@solana/web3.js';
import {
    TOKEN_PROGRAM_ID,
    ASSOCIATED_TOKEN_PROGRAM_ID,
} from '@solana/spl-token';
import { Ctx } from "../ctx/ctx";

export async function initializeRPC(ctx: Ctx) {
    await ctx.program.methods.initialize(ctx.owner.publicKey)
    .accounts({
        factory: ctx.PDAS.factory.key,
        rewardTokenMint: ctx.PDAS.factory.rewardTokenMint,
        stakeTokenMint: ctx.PDAS.factory.stakeTokenMint,
        vaultReward: ctx.PDAS.factory.vaultReward,
        initializer: ctx.owner.publicKey,
        systemProgram: SystemProgram.programId,
        tokenProgram: TOKEN_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        rent: anchor.web3.SYSVAR_RENT_PUBKEY,
    })
    .signers([ctx.owner])
    .rpc();
}

File: programs/staking/src/context/initialize.rs
use anchor_lang::prelude::*;
use crate::account::*;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token::{TokenAccount, Token, Mint};

#[derive(Accounts)]
pub struct Initialize<'info> {
    #[account(
        init, 
        payer = initializer,
        space = 8 + Factory::SPACE,
        seeds = [Factory::PDA_SEED], 
        bump,
    )]
    pub factory: Account<'info, Factory>,
    pub reward_token_mint: Box<Account<'info, Mint>>,
    pub stake_token_mint: Box<Account<'info, Mint>>,
    #[account(
        init,
        payer = initializer,
        associated_token::mint = reward_token_mint,
        associated_token::authority = factory,
    )]
    pub vault_reward: Box<Account<'info, TokenAccount>>,
    #[account(mut)]
    pub initializer: Signer<'info>,
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}
